### PR TITLE
Pv2-2592 ei redesign

### PR DESCRIPTION
**Pull request by hardik-desai-kcl**

### CODE REVIEW

**Review by AkshayBanifatsyevichChugh — DISMISSED** [review]

File: src/SFA.DAS.Payments.MatchedLearner.Api/Properties/launchSettings.json [hunk 1]
```diff
@@ -13,15 +13,15 @@
       "commandName": "IISExpress",
       "launchBrowser": true,
       "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development"
+        "EnvironmentName": "Development"
```
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.MatchedLearner.Api/Properties/launchSettings.json:16]: this variable is only used at Local development time so no need to change

File: src/SFA.DAS.Payments.MatchedLearner.Data/Contexts/MatchedLearnerDataContext.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            await Database.ExecuteSqlRawAsync($"DELETE FROM Payments2.DataLockEvent WHERE ukprn = @ukprn AND AcademicYear = @academicYear AND CollectionPeriod IN ({ sqlParamName })", sqlParameters);
+        }
+
+        public async Task RemoveApprenticeships(IEnumerable<long> apprenticeshipIds)
+        {
+            var sqlParameters = apprenticeshipIds.Select((item, index) => new SqlParameter($"@Id{index}", item)).ToList();
+            var sqlParamName = string.Join(", ", sqlParameters.Select(pn => pn.ParameterName));
+
+            await Database.ExecuteSqlRawAsync($"DELETE FROM Payments2.Apprenticeship WHERE id IN ( {sqlParamName} )", sqlParameters);
+        }
```
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.MatchedLearner.Data/Contexts/MatchedLearnerDataContext.cs:51]: these two methods deletes existing data from Matched Learner DB just before latest submission data is imported

File: src/SFA.DAS.Payments.MatchedLearner.Application/MatchedLearnerDataImportService.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        public MatchedLearnerDataImportService(IMatchedLearnerRepository matchedLearnerRepository, IPaymentsRepository paymentsRepository)
+        {
+            _matchedLearnerRepository = matchedLearnerRepository ?? throw new ArgumentNullException(nameof(matchedLearnerRepository));
+            _paymentsRepository = paymentsRepository ?? throw new ArgumentNullException(nameof(paymentsRepository));
+        }
+
+        public async Task Import(long ukprn, byte collectionPeriod, short academicYear)
+        {
+            //if current collection is 1 then only try remove current collection data otherwise remove current + previous collection data
```
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.MatchedLearner.Application/MatchedLearnerDataImportService.cs:28]: removed

File: src/SFA.DAS.Payments.MatchedLearner.Data/Entities/ApprenticeshipModel.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        public long? TransferSendingEmployerAccountId { get; set; }
+        public DateTime? StopDate { get; set; }
+        public int Priority { get; set; }
+        public ApprenticeshipStatus Status { get; set; }
+        public bool IsLevyPayer { get; set; }
+
+        // public List<ApprenticeshipPriceEpisodeModel> ApprenticeshipPriceEpisodes { get; set; }
+        public byte ApprenticeshipEmployerType { get; set; }
+
+        // public List<ApprenticeshipPauseModel> ApprenticeshipPauses { get; set; }
```
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.MatchedLearner.Data/Entities/ApprenticeshipModel.cs:29]: removed

File: src/SFA.DAS.Payments.MatchedLearner.Data/Contexts/MatchedLearnerDataContext.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            modelBuilder.ApplyConfiguration(new DataLockEventNonPayablePeriodFailureModelConfiguration());
+            modelBuilder.ApplyConfiguration(new DataLockEventNonPayablePeriodModelConfiguration());
+            modelBuilder.ApplyConfiguration(new DataLockEventPayablePeriodModelConfiguration());
+            modelBuilder.ApplyConfiguration(new DataLockEventPriceEpisodeModelConfiguration());
+        }
+
+        public async Task<int> SaveChanges(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await SaveChangesAsync(cancellationToken).ConfigureAwait(false);
+        }
```
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.MatchedLearner.Data/Contexts/MatchedLearnerDataContext.cs:70]: removed

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

File: src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
+﻿using System;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Logging;
+
+namespace SFA.DAS.Payments.MatchedLearner.Data
```
Comment (AkshayBanifatsyevichChugh) [inline src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs:10]: 3 classes and 1 interface in 1 file, hard to read

File: src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    public interface ISqlAzureIdentityTokenProvider
+    {
+        Task<string> GetAccessTokenAsync(CancellationToken cancellationToken = default);
+        string GetAccessToken();
+    }
+
+    public class SqlAzureIdentityTokenProvider : ISqlAzureIdentityTokenProvider
+    {
+        private readonly ILogger<SqlAzureIdentityTokenProvider> _logger;
+        private static readonly Microsoft.Azure.Services.AppAuthentication.AzureServiceTokenProvider AzureServiceTokenProvider = new Microsoft.Azure.Services.AppAuthentication.AzureServiceTokenProvider();
```
Comment (AkshayBanifatsyevichChugh) [inline src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs:21]: Few things:

- fully qualified type name used but not needed
- variable marked as static but access is private and class is non static, not sure if intended
- doesn't follow underscore convention
- if will be changed to non static, should move assignment of variable to constructor

File: src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        public string GetAccessToken()
+        {
+            var token = AzureServiceTokenProvider.GetAccessTokenAsync("https://database.windows.net/").GetAwaiter().GetResult();
+            
+            _logger.LogInformation("Generated SQL AccessToken");
+
+            return token;
+        }
```
Comment (AkshayBanifatsyevichChugh) [inline src/SFA.DAS.Payments.MatchedLearner.Data/SqlAzureIdentityTokenProvider.cs:44]: network call done synchronously

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

File: src/SFA.DAS.Payments.MatchedLearner.Application.UnitTests/RepositoryTests/MatchedLearnerRepositoryTests/WhenGettingDataLockEvents.cs [hunk 9]
```diff
@@ -173,6 +180,7 @@ private async Task AddDataLockToDb()
         {
             _dataLockEvent.LearnerUln = _uln;
             _dataLockEvent.Ukprn = _ukprn;
+            _dataLockEvent.EventId = Guid.NewGuid();
             _dataLockEvent.LearningAimReference = "ZPROG001";
             _dataLockEvent.CollectionPeriod = 11;
             _dataLockEvent.AcademicYear = 2021;
```
Comment (tom-gough) [inline src/SFA.DAS.Payments.MatchedLearner.Application.UnitTests/RepositoryTests/MatchedLearnerRepositoryTests/WhenGettingDataLockEvents.cs:186]: some commented out code in this class

**Review by tom-gough — CHANGES_REQUESTED** [review]

**Review by tom-gough — APPROVED** [review]

**Review by tom-gough — APPROVED** [review]

**Review by AkshayBanifatsyevichChugh — DISMISSED** [review]

**Review by tom-gough — APPROVED** [review]

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

**Review by AkshayBanifatsyevichChugh — DISMISSED** [review]

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

**Review by AkshayBanifatsyevichChugh — DISMISSED** [review]

**Review by AkshayBanifatsyevichChugh — APPROVED** [review]

**Pull request merged**